### PR TITLE
Add IOperation unit tests for try/catch statements and make the APIs …

### LINKED ISSUE #22008: API review for ITryStatement
https://github.com/dotnet/roslyn/blob/features/ioperation/src/Compilers/Core/Portable/Operations/ITryStatement.cs

Add unit tests when making the APIs public

### PR DESCRIPTION
**Pull request by mavasani**
…public again

Fixes #22008 and #21277

We also decided to make some changes to ICatchClause see https://github.com/dotnet/roslyn/issues/21277#issuecomment-330663344 and https://github.com/dotnet/roslyn/issues/21277#issuecomment-331310258

### CODE REVIEW

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_TryCatch.cs [hunk 1]
```diff
@@ -116,5 +120,638 @@ void M(int i)
...
+                //             int i = 0;
+                Diagnostic(ErrorCode.WRN_UnreferencedVarAssg, "i").WithArguments("i").WithLocation(13, 17)
+            };
+
+            VerifyOperationTreeAndDiagnosticsForTest<TryStatementSyntax>(source, expectedOperationTree, expectedDiagnostics);
+        }
+
+        [CompilerTrait(CompilerFeature.IOperation)]
+        [Fact]
+        public void TryCatch_CatchFilterWithLocalDeclaration()
```
Comment (mavasani) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_TryCatch.cs]: Locals are currently not exposed in the operation tree. https://github.com/dotnet/roslyn/issues/22301 tracks adding this API.

**Comment by mavasani** [comment]
This PR is now ready for review.

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs [hunk 2]
```diff
@@ -304,8 +304,13 @@ private IInvocationExpression CreateBoundCallOperation(BoundCall boundCall)
             return new LazyInvocationExpression(targetMethod, instance, isVirtual, argumentsInEvaluationOrder, _semanticModel, syntax, type, constantValue, isImplicit);
         }
 
-        private ILocalReferenceExpression CreateBoundLocalOperation(BoundLocal boundLocal)
+        private IOperation CreateBoundLocalOperation(BoundLocal boundLocal)
         {
+            if (boundLocal.Syntax.Kind() == SyntaxKind.CatchDeclaration)
+            {
+                return CreateVariableDeclaration(boundLocal);
+            }
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: Why not have `CreateBoundCatchBlockOperation` call `CreateVariableDeclaration()` directly rather than handling declarations here?

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs [hunk 3]
```diff
@@ -1284,15 +1288,15 @@ private ITryStatement CreateBoundTryStatementOperation(BoundTryStatement boundTr
 
         private ICatchClause CreateBoundCatchBlockOperation(BoundCatchBlock boundCatchBlock)
         {
-            Lazy<IBlockStatement> handler = new Lazy<IBlockStatement>(() => (IBlockStatement)Create(boundCatchBlock.Body));
-            ITypeSymbol caughtType = boundCatchBlock.ExceptionTypeOpt;
+            Lazy<IOperation> expressionDeclarationOrExpression = new Lazy<IOperation>(() => boundCatchBlock.ExceptionSourceOpt != null ? Create(boundCatchBlock.ExceptionSourceOpt) : null);
+            ITypeSymbol exceptionType = boundCatchBlock.ExceptionTypeOpt ?? ((CSharpCompilation)_semanticModel.Compilation).GetWellKnownType(WellKnownType.System_Exception);
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: If `ExceptionTypeOpt` is `null`, we cannot assume the exception type is `Exception`.

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs (hunk 2 above)
Comment (mavasani) [reply src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: We need to go through Create code path to ensure the bound node gets added to the bound node to operation cache.

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs (hunk 3 above)
Comment (mavasani) [reply src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: I should probably change this check to see if the catch `Declaration` is null, then default to `System.Exception`, otherwise use `ExceptionTypeOpt`

File: src/Compilers/Core/Portable/Operations/ITryStatement.cs [hunk 4]
```diff
@@ -11,7 +11,7 @@ namespace Microsoft.CodeAnalysis.Semantics
     /// This interface is reserved for implementation by its associated APIs. We reserve the right to
     /// change it in the future.
     /// </remarks>
-    internal interface ITryStatement : IOperation // https://github.com/dotnet/roslyn/issues/22008
+    public interface ITryStatement : IOperation
```
Comment (cston) [inline src/Compilers/Core/Portable/Operations/ITryStatement.cs]: Consider naming `FinallyHandler` as `Finally` since we don't appear to use the term `Handler` for `Catches`.

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs (hunk 3 above)
Comment (cston) [reply src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: If the declaration is missing, the type should be `null`.

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: > If the declaration is missing, the type should be null.

I agree with this, It looks like C# emits a catch that can catch any type in this case.

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs [hunk 5]
```diff
@@ -304,8 +304,13 @@ private IInvocationExpression CreateBoundCallOperation(BoundCall boundCall)
             return new LazyInvocationExpression(targetMethod, instance, isVirtual, argumentsInEvaluationOrder, _semanticModel, syntax, type, constantValue, isImplicit);
         }
 
-        private ILocalReferenceExpression CreateBoundLocalOperation(BoundLocal boundLocal)
+        private IOperation CreateBoundLocalOperation(BoundLocal boundLocal)
         {
+            if (boundLocal.Syntax.Kind() == SyntaxKind.CatchDeclaration)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: >if (boundLocal.Syntax.Kind() == SyntaxKind.CatchDeclaration) [](start = 12, length = 60)

I think it would be better to have this special case in code that deals with BoundCatchBlock rather than making this function context aware.

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs [hunk 6]
```diff
@@ -1284,15 +1288,15 @@ private ITryStatement CreateBoundTryStatementOperation(BoundTryStatement boundTr
...
+            ITypeSymbol exceptionType = boundCatchBlock.ExceptionTypeOpt ?? ((CSharpCompilation)_semanticModel.Compilation).GetWellKnownType(WellKnownType.System_Exception);
             Lazy<IOperation> filter = new Lazy<IOperation>(() => Create(boundCatchBlock.ExceptionFilterOpt));
-            ILocalSymbol exceptionLocal = (boundCatchBlock.Locals.FirstOrDefault()?.DeclarationKind == CSharp.Symbols.LocalDeclarationKind.CatchVariable) ? boundCatchBlock.Locals.FirstOrDefault() : null;
+            Lazy<IBlockStatement> handler = new Lazy<IBlockStatement>(() => (IBlockStatement)Create(boundCatchBlock.Body));
             SyntaxNode syntax = boundCatchBlock.Syntax;
             ITypeSymbol type = null;
             Optional<object> constantValue = default(Optional<object>);
             bool isImplicit = boundCatchBlock.WasCompilerGenerated;
-            return new LazyCatchClause(handler, caughtType, filter, exceptionLocal, _semanticModel, syntax, type, constantValue, isImplicit);
+            return new LazyCatchClause(expressionDeclarationOrExpression, exceptionType, filter, handler, _semanticModel, syntax, type, constantValue, isImplicit);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: >return new LazyCatchClause(expressionDeclarationOrExpression, exceptionType, filter, handler, _semanticModel, syntax, type, constantValue, isImplicit); [](start = 12, length = 151)

It doesn't look like we are exposing ```boundCatchBlock.Locals```.

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_TryCatch.cs [hunk 7]
```diff
@@ -116,5 +120,638 @@ void M(int i)
...
+        {
+        }
+        catch (Exception) /*<bind>*/when (s != null)/*</bind>*/
+        {
+        }
+    }
+}
+";
+            string expectedOperationTree = @"
+Filter: IBinaryOperatorExpression (BinaryOperatorKind.NotEquals) (OperationKind.BinaryOperatorExpression, Type: System.Boolean) (Syntax: 's != null')
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_TryCatch.cs]: I would expect to get nothing here, there is no node associated with CatchFilterClauseSyntax.

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_TryCatch.cs [hunk 8]
```diff
@@ -116,5 +120,638 @@ void M(int i)
...
+        }
+        /*<bind>*/finally
+        {
+            Console.WriteLine(s);
+        }/*</bind>*/
+    }
+}
+";
+            string expectedOperationTree = @"
+Finally: IBlockStatement (1 statements) (OperationKind.BlockStatement) (Syntax: '{ ... }')
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_TryCatch.cs]: I would expect to get nothing here, there is no node associated with FinallyClauseSyntax

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_TryCatch.cs [hunk 9]
```diff
@@ -116,5 +120,638 @@ void M(int i)
...
+using System;
+
+class C
+{
+    static void M(string s)
+    {
+        try
+        {
+        }
+        catch (Exception) /*<bind>*/when (s != null)/*</bind>*/
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_TryCatch.cs]: >s != null [](start = 42, length = 9)

It would be good to have a test that requests node for CatchFilterClauseSyntax.FilterExpression

File: src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb [hunk 10]
```diff
@@ -799,7 +799,11 @@ Namespace Microsoft.CodeAnalysis.Semantics
             Return New ParameterReferenceExpression(parameter, _semanticModel, syntax, type, constantValue, isImplicit)
         End Function
 
-        Private Function CreateBoundLocalOperation(boundLocal As BoundLocal) As ILocalReferenceExpression
+        Private Function CreateBoundLocalOperation(boundLocal As BoundLocal) As IOperation
+            If boundLocal.Syntax.Kind = SyntaxKind.IdentifierName AndAlso boundLocal.Syntax.Parent?.Kind = SyntaxKind.CatchStatement Then
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb]: >If boundLocal.Syntax.Kind = SyntaxKind.IdentifierName AndAlso boundLocal.Syntax.Parent?.Kind = SyntaxKind.CatchStatement Then [](start = 12, length = 125)

Same comment as for C# implementation.

File: src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb [hunk 11]
```diff
@@ -1077,15 +1081,16 @@ Namespace Microsoft.CodeAnalysis.Semantics
         End Function
 
         Private Function CreateBoundCatchBlockOperation(boundCatchBlock As BoundCatchBlock) As ICatchClause
-            Dim handler As Lazy(Of IBlockStatement) = New Lazy(Of IBlockStatement)(Function() DirectCast(Create(boundCatchBlock.Body), IBlockStatement))
-            Dim caughtType As ITypeSymbol = boundCatchBlock.ExceptionSourceOpt?.Type
+            Dim exceptionDeclarationOrExpression As Lazy(Of IOperation) = New Lazy(Of IOperation)(
+                Function() If(boundCatchBlock.ExceptionSourceOpt IsNot Nothing, Create(boundCatchBlock.ExceptionSourceOpt), Nothing))
+            Dim exceptionType As ITypeSymbol = If(boundCatchBlock.ExceptionSourceOpt?.Type, DirectCast(_semanticModel.Compilation, VisualBasicCompilation).GetWellKnownType(WellKnownType.System_Exception))
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb]: >Dim exceptionType As ITypeSymbol = If(boundCatchBlock.ExceptionSourceOpt?.Type, DirectCast(_semanticModel.Compilation, VisualBasicCompilation).GetWellKnownType(WellKnownType.System_Exception)) [](start = 12, length = 192)

This is correct for VB.

File: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_TryCatch.vb [hunk 12]
```diff
@@ -104,5 +110,742 @@ IBlockStatement (3 statements) (OperationKind.BlockStatement) (Syntax: 'Private
...
+        End Try
+    End Sub
+End Class]]>.Value
+
+            Dim expectedOperationTree = <![CDATA[
+ITryStatement (OperationKind.TryStatement) (Syntax: 'Try'BIND:"T ... End Try')
+  Body: IBlockStatement (0 statements) (OperationKind.BlockStatement) (Syntax: 'Try'BIND:"T ... End Try')
+  Catch clauses(1):
+      ICatchClause (Exception type: System.Exception) (OperationKind.CatchClause) (Syntax: 'Catch e')
+        ExceptionDeclarationOrExpression: IVariableDeclaration (1 variables) (OperationKind.VariableDeclaration) (Syntax: 'e')
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_TryCatch.vb]: >ExceptionDeclarationOrExpression: IVariableDeclaration (1 variables) (OperationKind.VariableDeclaration) (Syntax: 'e') [](start = 8, length = 118)

This is wrong, the variable is declared outside of the catch.

File: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_TryCatch.vb [hunk 13]
```diff
@@ -104,5 +110,742 @@ IBlockStatement (3 statements) (OperationKind.BlockStatement) (Syntax: 'Private
...
+        End Sub
+
+        <CompilerTrait(CompilerFeature.IOperation)>
+        <Fact>
+        Public Sub TryCatch_CatchClauseWithLocalReferenceExpression()
+            Dim source = <![CDATA[
+Imports System
+Class C
+    Private Shared Sub M()
+        Dim e As Exception = Nothing
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_TryCatch.vb]: >Exception [](start = 17, length = 9)

Consider testing with a different exception type.

File: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_TryCatch.vb [hunk 14]
```diff
@@ -104,5 +110,742 @@ IBlockStatement (3 statements) (OperationKind.BlockStatement) (Syntax: 'Private
...
+            VerifyOperationTreeAndDiagnosticsForTest(Of TryBlockSyntax)(source, expectedOperationTree, expectedDiagnostics)
+        End Sub
+
+        <CompilerTrait(CompilerFeature.IOperation)>
+        <Fact>
+        Public Sub TryCatch_CatchClauseWithParameterReferenceExpression()
+            Dim source = <![CDATA[
+Imports System
+Class C
+    Private Shared Sub M(e As Exception)
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_TryCatch.vb]: >Exception [](start = 30, length = 9)

Consider testing with a different exception type.

File: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_TryCatch.vb [hunk 15]
```diff
@@ -104,5 +110,742 @@ IBlockStatement (3 statements) (OperationKind.BlockStatement) (Syntax: 'Private
...
+            VerifyOperationTreeAndDiagnosticsForTest(Of TryBlockSyntax)(source, expectedOperationTree, expectedDiagnostics)
+        End Sub
+
+        <CompilerTrait(CompilerFeature.IOperation)>
+        <Fact>
+        Public Sub TryCatch_CatchClauseWithFieldReferenceExpression()
+            Dim source = <![CDATA[
+Imports System
+Class C
+    Private e As Exception = Nothing
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_TryCatch.vb]: >Exception [](start = 17, length = 9)

Consider testing with a different exception type.

File: src/Compilers/Core/Portable/Operations/ICatchClause.cs [hunk 16]
```diff
@@ -11,24 +9,31 @@ namespace Microsoft.CodeAnalysis.Semantics
...
         /// <summary>
-        /// Type of exception to be handled.
+        /// Type of the exception handled by the catch clause.
         /// </summary>
-        ITypeSymbol CaughtType { get; }
+        ITypeSymbol ExceptionType { get; }
+
         /// <summary>
-        /// Filter expression to be executed to determine whether to handle the exception.
+        /// Optional source for exception. This could be any of the following operation:
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Operations/ICatchClause.cs]: >Optional source for exception. This could be any of the following operation: [](start = 12, length = 76)

It looks like the list is not complete. For example, one of the tests have a field reference expression as the ```ExceptionDeclarationOrExpression```. Consider rephrasing item 4: "Other expression for error scenarios in VB"

File: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_TryCatch.vb [hunk 17]
```diff
@@ -104,5 +110,742 @@ IBlockStatement (3 statements) (OperationKind.BlockStatement) (Syntax: 'Private
...
+    Private Shared Sub M(s As String)
+        Try
+        Finally'BIND:"Finally"
+            Console.WriteLine(s)
+        End Try
+    End Sub
+End Class]]>.Value
+
+            ' GetOperation returns Nothing for FinallyStatementSyntax
+            Assert.Null(GetOperationTreeForTest(Of FinallyStatementSyntax)(source).operation)
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_TryCatch.vb]: >GetOperationTreeForTest [](start = 24, length = 23)

It would be good to test what nodes do we get for EndTryStatement, CatchFilterClause, CatchFilterClause.Filter and for an individual statement inside Try/Catch/Finally blocks.

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 1)

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs (hunk 2 above)
Comment (mavasani) [reply src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: Actually, sounds like this is a better suggestion given our new GetOperation approach. Let me fix this.

---
In reply to: [140611924](https://github.com/dotnet/roslyn/pull/22210#discussion_r140611924) [](ancestors = 140611924)

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs (hunk 6 above)
Comment (mavasani) [reply src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: These exposed through the underlying IBlockStatement of the handler. See [https://github.com/dotnet/roslyn/pull/22210/files#diff-86f74ac187f645e9c3f76409a6fac779R467](https://github.com/dotnet/roslyn/pull/22210/files#diff-86f74ac187f645e9c3f76409a6fac779R467)

---
In reply to: [141126304](https://github.com/dotnet/roslyn/pull/22210#discussion_r141126304) [](ancestors = 141126304)

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_TryCatch.cs (hunk 7 above)
Comment (mavasani) [reply src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_TryCatch.cs]: Currently, we do return null, so I had skipped this test with [https://github.com/dotnet/roslyn/issues/22299](https://github.com/dotnet/roslyn/issues/22299). I will close that issue and update this test to verify null.

---
In reply to: [141130773](https://github.com/dotnet/roslyn/pull/22210#discussion_r141130773) [](ancestors = 141130773)

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_TryCatch.cs (hunk 8 above)
Comment (mavasani) [reply src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_TryCatch.cs]: Currently, we do return null, so I had skipped this test with [https://github.com/dotnet/roslyn/issues/22299](https://github.com/dotnet/roslyn/issues/22299). I will close that issue and update this test to verify null.

---
In reply to: [141131397](https://github.com/dotnet/roslyn/pull/22210#discussion_r141131397) [](ancestors = 141131397)

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_TryCatch.cs (hunk 9 above)
Comment (mavasani) [reply src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_TryCatch.cs]: Added `TryCatch_GetOperationForCatchFilterClauseExpression`

---
In reply to: [141132435](https://github.com/dotnet/roslyn/pull/22210#discussion_r141132435) [](ancestors = 141132435)

**Review by cston — APPROVED** [review]

**Comment by mavasani** [comment]
@dotnet-bot retest windows_debug_unit64_prtest please

**Comment by mavasani** [comment]
@AlekseyTs - can you please take another look?

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs (hunk 6 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: > These exposed through the underlying IBlockStatement of the handler. See [https://github.com/dotnet/roslyn/pull/22210/files#diff-86f74ac187f645e9c3f76409a6fac779R467](https://github.com/dotnet/roslyn/pull/22210/files#diff-86f74ac187f645e9c3f76409a6fac779R467)

I don't think these are the locals I am talking about. I am talking about the exception local and the locals declared in the filter. 

---
In reply to: [141348394](https://github.com/dotnet/roslyn/pull/22210#discussion_r141348394) [](ancestors = 141348394,141126304)

File: src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb [hunk 18]
```diff
@@ -695,13 +691,14 @@ Namespace Microsoft.CodeAnalysis.Semantics
         End Function
 
         Private Function CreateBoundArrayCreationOperation(boundArrayCreation As BoundArrayCreation) As IArrayCreationExpression
+            Dim elementType As ITypeSymbol = TryCast(boundArrayCreation.Type, IArrayTypeSymbol)?.ElementType
             Dim dimensionSizes As Lazy(Of ImmutableArray(Of IOperation)) = New Lazy(Of ImmutableArray(Of IOperation))(Function() boundArrayCreation.Bounds.SelectAsArray(Function(n) Create(n)))
             Dim initializer As Lazy(Of IArrayInitializer) = New Lazy(Of IArrayInitializer)(Function() DirectCast(Create(boundArrayCreation.InitializerOpt), IArrayInitializer))
             Dim syntax As SyntaxNode = boundArrayCreation.Syntax
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb]: >DirectCast(boundCatchBlock.ExceptionSourceOpt, BoundLocal) [](start = 49, length = 58)

We probably should not assume that this cast is safe, adding the kind check to the ```If``` should take care of that.

File: src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb [hunk 19]
```diff
@@ -695,13 +691,14 @@ Namespace Microsoft.CodeAnalysis.Semantics
...
             Dim type As ITypeSymbol = boundArrayCreation.Type
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb]: >boundCatchBlock.LocalOpt Is exceptionSourceOpt.LocalSymbol [](start = 37, length = 58)

I believe we should add this condition to the ```If``` instead of asserting it.

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs (hunk 6 above)
Comment (mavasani) [reply src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: Ah, did not realize that. I had filed [https://github.com/dotnet/roslyn/issues/22301](https://github.com/dotnet/roslyn/issues/22301) to track this work, but the API is straightforward so I have added it to this PR.

---
In reply to: [141764536](https://github.com/dotnet/roslyn/pull/22210#discussion_r141764536) [](ancestors = 141764536,141348394,141126304)

**Comment by mavasani** [comment]
@AlekseyTs - Addressed your feedback.

Tagging @jcouv for ask mode approval.

**Review by AlekseyTs — APPROVED** [review]
LGTM (iteration 4)

**Comment by mavasani** [comment]
@dotnet-bot retest windows_debug_unit64_prtest please

**Comment by mavasani** [comment]
@dotnet-bot retest windows_debug_unit64_prtest please

**Comment by mavasani** [comment]
@dotnet-bot retest windows_debug_unit64_prtest please

File: src/Compilers/Core/Portable/Operations/ICatchClause.cs [hunk 20]
```diff
@@ -11,24 +11,36 @@ namespace Microsoft.CodeAnalysis.Semantics
...
-        IOperation Filter { get; }
+        ITypeSymbol ExceptionType { get; }
+
         /// <summary>
-        /// Symbol for the local catch variable bound to the caught exception.
+        /// Optional source for exception. This could be any of the following operation:
+        /// 1. Declaration for the local catch variable bound to the caught exception (C# and VB) OR
+        /// 2. Null, indicating no declaration or expression (C# and VB)
+        /// 3. Reference to an existing local or parameter (VB) OR
+        /// 4. Other expression for error scenarios (VB)
```
Comment (333fred) [inline src/Compilers/Core/Portable/Operations/ICatchClause.cs]: This list and the list in Operations.xml.Generated.cs are not in sync.

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs [hunk 21]
```diff
@@ -304,7 +304,7 @@ private IInvocationExpression CreateBoundCallOperation(BoundCall boundCall)
             return new LazyInvocationExpression(targetMethod, instance, isVirtual, argumentsInEvaluationOrder, _semanticModel, syntax, type, constantValue, isImplicit);
         }
 
-        private ILocalReferenceExpression CreateBoundLocalOperation(BoundLocal boundLocal)
+        private IOperation CreateBoundLocalOperation(BoundLocal boundLocal)
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: Unrelated?

File: src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb [hunk 22]
```diff
@@ -818,7 +818,7 @@ Namespace Microsoft.CodeAnalysis.Semantics
             Return New ParameterReferenceExpression(parameter, _semanticModel, syntax, type, constantValue, isImplicit)
         End Function
 
-        Private Function CreateBoundLocalOperation(boundLocal As BoundLocal) As ILocalReferenceExpression
+        Private Function CreateBoundLocalOperation(boundLocal As BoundLocal) As IOperation
```
Comment (333fred) [inline src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb]: Unrelated?

**Review by 333fred — APPROVED** [review]
LGTM, aside from my small comments.

**Comment by mavasani** [comment]
Thanks Fred, I will handle your feedback separately in next PR
